### ISSUE TITLE
Gin master build failed

### ISSUE THREAD

**Issue by chainhelen** [gin-gonic/gin#1511]
@thinkerou @appleboy 
Gin master, the commit is [0da5b0c85ac2539e775a7e63aebf3e0b5808dab4](https://github.com/gin-gonic/gin/tree/0da5b0c85ac2539e775a7e63aebf3e0b5808dab4)

## Description
refercens
#1510  test(not review): add go1.11 
#1383 Support map as query string or post form parameters  
#1362 How can I get map[string]string filed on PostForm?
#1106 Map as query parameter

Build failed in go master but ok in other go version.
![image](https://user-images.githubusercontent.com/7046329/44564093-d88dc180-a793-11e8-9408-17e03e745bd5.png)

## Reason 
This  because this [commit](https://github.com/golang/go/commit/1040626c0ce4a1bc2b312aa0866ffeb2ff53c1ab#diff-6c2d018290e298803c0c9419d8739885R1124) in go master, add this [func validQuery](https://github.com/golang/go/commit/1040626c0ce4a1bc2b312aa0866ffeb2ff53c1ab#diff-6c2d018290e298803c0c9419d8739885R1124)
 ```
// validQuery reports whether s is a valid query string per RFC 3986
// Section 3.4:
//     query       = *( pchar / "/" / "?" )
//     pchar       = unreserved / pct-encoded / sub-delims / ":" / "@"
//     unreserved  = ALPHA / DIGIT / "-" / "." / "_" / "~"
//     sub-delims  = "!" / "$" / "&" / "'" / "(" / ")"
//                   / "*" / "+" / "," / ";" / "="
func validQuery(s string) bool {
	pctEnc := 0
 	for _, r := range s {
		if pctEnc > 0 {
			if uint32(r) > 255 || !ishex(byte(r)) {
				return false
			}
			pctEnc--
			continue
		} else if r == '%' {
			pctEnc = 2
			continue
		}
 		if 'A' <= r && r <= 'Z' {
			continue
		}
		if 'a' <= r && r <= 'z' {
			continue
		}
		if '0' <= r && r <= '9' {
			continue
		}
		switch r {
		case '-', '.', '_', '~', '!', '$', '&', '\'', '(', ')',
			'*', '+', ',', ';', '=', ':', '@', '/', '?':
			continue
		default:
			return false
		}
	}
 	return true
}
```  
Then I copy these into local workspace,  test like below 
```
func main() {
    fmt.Printf("%t\n", validQuery("both=GET&id=main&id=omit&array[]=first&array[]=second&ids[a]=hi&ids[b]=3.14"))
    fmt.Printf("%t\n", validQuery("foo=bar&page=10&id=2"))
}

// build and run, get 
false
true
``` 

## Summarize
In go master, if url include '[' or ']', will be [QueryEscape](https://github.com/golang/go/commit/1040626c0ce4a1bc2b312aa0866ffeb2ff53c1ab#diff-6c2d018290e298803c0c9419d8739885R523) according to `RFC 3986`. So don't parse url.query

**Comment by chainhelen** [comment]
Maybe we can parse this type query in gin itself.

**Comment by thinkerou** [comment]
@chainhelen fix issue, please see #1510 and help review, thanks!
